File: my-folder/0112-path-sum/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def traverse(self, root, targetSum):
        if not root: return
        
        #leaf
        if (not root.left and not root.right) and targetSum-root.val == 0:
            return True
    
        left = self.traverse(root.left, targetSum-root.val)
        right = self.traverse(root.right, targetSum-root.val)
        
        return left or right
    
    def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:
        if not root: return False
        
        return self.traverse(root, targetSum)

File: my-folder/0112-path-sum/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */

var traverse = function(root, targetSum) {
    if (root === null) return false
    
    if (root.left === null && root.right === null && targetSum-root.val===0) {
        return true
    }
    
    var left = traverse(root.left, targetSum-root.val)
    var right = traverse(root.right, targetSum-root.val)
    return left || right
    
}

var hasPathSum = function(root, targetSum) {
    if (root === null) return false
    return traverse(root, targetSum)
};
